### ISSUE TITLE
Technical Ideas

### ISSUE THREAD

**Issue by dcgoss** [cognoma/cognoma#4]
I know [Google Cloud](https://cloud.google.com) has some really excellent machine learning offerings. It seems [Google Genomics](https://cloud.google.com/genomics/) and [Google Cloud Machine Learning](https://cloud.google.com/ml/) may come in handy for infrastructure and data.

**Comment by awm33** [comment]
Google Cloud Platform is an excellent and I really like BigQuery. The project should probably stick to non-proprietary technology, someone should be able to run this entirely on their own servers if they need to. Some Google technology like Tensor Flow and Apache Beam (Dataflow) are open source. They also offer hosted Hadoop and Spark, so does AWS.

**Comment by dcgoss** [comment]
Agreed, and I can also vouch for Google Cloud.  I think hosted services are fine for us, provided that the underlying service is open source. One of the key points of Google Cloud ML is its basically just hosted TensorFlow - you could do the same things it does, but GCML just does it for you on their own managed infrastructure.

**Comment by dcgoss** [comment]
On the backend it looks like we're going to need a web tier (Django?), a RDBMS tier (Postgres?), and a machine learning/data processing tier (need to figure out what types of models we need; Tensorflow? SKLearn?). Planning for scalability (load balancing, automated provisioning, docker?) from the start is prudent, but that piece doesn't need to actually be in place until it's needed and the rest of the service is ready.

**Comment by awm33** [comment]
I assume the lab team wants to create a single django app. Given the complex data backend, it could benefit from a microservice architecture, but that may be too complex for this project

I would place a job management system between the app and the data cluster management system (Hadoop, Spark, Tensorflow, etc). I've used a bunch of different kinds, but a MySQL/Postgres table plus queue (like SQS) works best. You could also use Redis or Zookeeper instead of a queue, something to lock the job. Could also use the job system for other things (sending emails, handling any background jobs).

I would recommend Postgres over MySQL. This would be for app data (users, jobs, data_sources, etc).

Choosing what to run the machine learning on top of would probably require input from data scientist and anyone working with ML code. If we stick with Python, Tensorflow or Scikit-learn on top of Hadoop/Spark seem to be the obvious choices. Could also run Scikit-learn on parent-child jobs if we want to skip Hadoop/Spark.

I agree with using Docker, it would make development easier and can be used for deployment on all the major cloud platforms plus private servers.

**Comment by dcgoss** [comment]
More ideas for job/task management: celery, gearman, rabbitmq

**Comment by cgreene** [comment]
We have used celery & rabbitmq in the past with django servers and it worked well.

I'd also love to see things containerized with Docker to facilitate deployment.

**Comment by awm33** [comment]
I've written two task queues / management services, work with another one at my company, and I've used a few other open source ones. The service I wrote and use now exposes a restful interface. Been thinking of open sourcing it or writing one for open source. The UI, code queuing tasks, and workers hit the REST interface directly.

I would say having a SQL table for job reporting has been incredibly valuable and the biggest thing I could recommend. For things like reporting and when things go wrong, being able to query using SQL rather than fumbling with Redis or whatever is operating the queue. This could be combined with a job queue lib as well, so if you're using celery, the queuer and worker would update the job reporting table.

**Comment by cgreene** [comment]
@awm33 Exactly - we have had the job responsible for those aspects. It sounds like you have much more recent experience in the area though - can you sketch out what you are thinking for this part either from scratch or starting from @dhimmel's draw.io? We'd love to take advantage of your experience with this to avoid making decisions we regret later!

**Comment by dhimmel** [comment]
Closing as future discussion can be directed to specific team repositories.
